File: 0567_PermutationInString.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    internal class _0567_PermutationInString
    {
        /*
        
        Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.

        In other words, return true if one of s1's permutations is the substring of s2.

        Example 1:
        Input: s1 = "ab", s2 = "eidbaooo"
        Output: true
        Explanation: s2 contains one permutation of s1 ("ba").
        
        Example 2:
        Input: s1 = "ab", s2 = "eidboaoo"
        Output: false

        Constraints:

        1 <= s1.length, s2.length <= 10^4
        s1 and s2 consist of lowercase English letters.

        */

        //var s1 = "ab";
        //var s2 = "eidbaooo";

        //var s1 = "ab";
        //var s2 = "eidboaoo";

        public bool CheckInclusion(string s1, string s2)
        {
            {
                if (s1.Length > s2.Length) return false;

                var s1Count = Enumerable.Repeat(0, 26).ToArray(); // Create arrays of 26 0's
                var s2Count = Enumerable.Repeat(0, 26).ToArray();

                // Instead of calculating each permutation then brute forcing, you can just create a hashmap
                // of the string that you are checking permutations of, then tally the counts of each present character
                for (var i = 0; i < s1.Length; i++)
                {
                    s1Count[s1[i] - 'a']++;
                    s2Count[s2[i] - 'a']++;
                }

                // Now perform a sliding window then iterate through each hashmap calculation.
                // If there is ever a completely equal hashmap comparison, then a true condition has been found

                var matches = 0;
                for (var i = 0; i < 26; i++)
                {
                    if (s1Count[i] == s2Count[i]) matches++;
                }

                var left = 0;
                for (var right = s1.Length; right < s2.Length; right++)
                {
                    if (matches == 26) return true;

                    var index = s2[right] - 'a';
                    s2Count[index]++;
                    if (s1Count[index] == s2Count[index])
                    {
                        matches++;
                    }
                    else if (s1Count[index] + 1 == s2Count[index])
                    {
                        matches--;
                    }

                    index = s2[left] - 'a';
                    s2Count[index]--;
                    if (s1Count[index] == s2Count[index])
                    {
                        matches++;
                    }
                    else if (s1Count[index] - 1 == s2Count[index])
                    {
                        matches--;
                    }

                    left++;
                }

                return matches == 26;
            }
        }

    }
}

File: C++ Solutions/LeetCodeCpp/0567_PermutationInString.cpp
/*

Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.

In other words, return true if one of s1's permutations is the substring of s2.

Example 1:
Input: s1 = "ab", s2 = "eidbaooo"
Output: true
Explanation: s2 contains one permutation of s1 ("ba").

Example 2:
Input: s1 = "ab", s2 = "eidboaoo"
Output: false

Constraints:

1 <= s1.length, s2.length <= 10^4
s1 and s2 consist of lowercase English letters.

*/

#include <vector>
#include <stack>
#include <string>
#include <unordered_set>

using namespace std;

class _0567_PermutationInString {
public:
    bool checkInclusion(string s1, string s2) {
        int m = s1.size();
        int n = s2.size();
        if (m > n) {
            return false;
        }

        vector<int> count(26);
        for (int i = 0; i < m; i++) {
            count[s1[i] - 'a']++;
            count[s2[i] - 'a']--;
        }
        if (isPermutation(count)) {
            return true;
        }

        for (int i = m; i < n; i++) {
            count[s2[i] - 'a']--;
            count[s2[i - m] - 'a']++;
            if (isPermutation(count)) {
                return true;
            }
        }

        return false;
    }
private:
    bool isPermutation(vector<int>& count) {
        for (int i = 0; i < 26; i++) {
            if (count[i] != 0) {
                return false;
            }
        }
        return true;
    }
};